drivers/xen/gntdev-dmabuf.c

```diff
@@ -88,6 +112,64 @@ struct gntdev_dmabuf_priv {
 
 static void dmabuf_exp_release(struct kref *kref);
 
+static struct gntdev_dmabuf_wait_obj *
+dmabuf_map_wait_obj_find(struct gntdev_dmabuf_priv *priv, int fd)
+{
+	struct gntdev_dmabuf_wait_obj *obj, *ret = ERR_PTR(-ENOENT);
+
+	mutex_lock(&priv->lock);
+	list_for_each_entry(obj, &priv->map_wait_list, next)
+		if (obj->fd == fd) {
+			pr_debug("Found gntdev_dmabuf in the wait list\n");
+			ret = obj;
+			break;
+		}
+	mutex_unlock(&priv->lock);
+
+	return ret;
+}
+
+static void dmabuf_exp_wait_obj_free(struct gntdev_dmabuf_priv *priv,
+			struct gntdev_dmabuf_wait_obj *obj);
+
+static int
+dmabuf_map_wait_obj_set(struct gntdev_dmabuf_priv *priv,
+			struct gntdev_dmabuf *gntdev_dmabuf, int fd)
+{
+	struct gntdev_dmabuf_wait_obj *obj;
+
+	obj = dmabuf_map_wait_obj_find(gntdev_dmabuf->priv, fd);
+	if ((!obj) || (PTR_ERR(obj) == -ENOENT)) {
+		obj = kzalloc(sizeof(*obj), GFP_KERNEL);
+		if (!obj)
+			return -ENOMEM;
+	}
+
+	init_completion(&obj->completion);
+	obj->gntdev_dmabuf = gntdev_dmabuf;
+	obj->fd = fd;
+	mutex_lock(&priv->lock);
+	list_add(&obj->next, &priv->map_wait_list);
+	mutex_unlock(&priv->lock);
+	return 0;
+}
+
+static void dmabuf_map_wait_obj_signal(struct gntdev_dmabuf_priv *priv,
+				       struct gntdev_dmabuf *gntdev_dmabuf)
+{
+	struct gntdev_dmabuf_wait_obj *obj;
+
+	mutex_lock(&priv->lock);
```

Comment:
Another tricky question: are you sure that there will be no race between call to this function and call to `gntdev_dmabuf_fini()`. Correct me, but `priv` might not exist at a moment, when this callback is called.

I don't see where you call `dma_buf_unregister_release_notifier`, BTW.

Reply:
Well, I am not sure how it supposed to work. `gntdev_dmabuf_fini` is called from `gntdev_release` which in turn is called when userspace closes `/dev/xen/gntdev`.  `gntdev_dmabuf_fini` causes call to `kfree(priv)` eventually. 

What if userpace issued `IOCTL_GNTDEV_DMABUF_MAP_RELEASE` and then quicky closed `/dev/xen/gntdev` ? You will have an active wait object and no `priv`. Also, what if userspace calls `IOCTL_GNTDEV_DMABUF_MAP_RELEASE` twice for the same buffer?

As for `dma_buf_unregister_release_notifier` - I believe you need to call it or there will be invalid entries in the chain. Why not call it from `dmabuf_release_notifier_cb` ?